source/agora/consensus/validation/Block.d

```diff
@@ -112,6 +118,268 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
     return null;
 }
 
+/*******************************************************************************
+
+    Check the validity of a genesis block
+
+    Follow the Block rules except for the following
+        - Block height is 0
+        - The previous block hash of the block must be empty
+        - The number of transactions in the block is above zero and less than
+          Block.TxsInBlock
+        - All the transactions pass validation, which implies:
+            - Transaction type in block should only be Payment or Freeze
+            - There is empty a input for the transaction
+            - There is one or more outputs of the transaction
+        - All the enrollments pass validation, which implies:
+            - The enrollments refer to freeze tx's in this block
+
+    Params:
+        block = The genesis block to check
+
+    Returns:
+        `null` if the genesis block is valid, a string explaining the reason it
+        is invalid otherwise.
+
+*******************************************************************************/
+
+private string isValidGenesisBlock (const ref Block genesisBlock) nothrow @safe
+{
+    if (genesisBlock.header.height != 0)
+        return "GenesisBlock: The height of the block is not 0";
+
+    if (genesisBlock.header.prev_block != Hash.init)
+        return "GenesisBlock: Header.prev_block is not empty";
+
+    if (genesisBlock.txs.length == 0)
+        return "GenesisBlock: Transaction(s) is empty";
+
+    if (!genesisBlock.txs.isSorted())
+        return "GenesisBlock: Transactions are not sorted";
+
+    if (genesisBlock.txs.length > Block.TxsInBlock)
+        return "GenesisBlock: The number of transactions is out of bounds";
```

Comment:
This right here is your opportunity for a DDoS. Consider that you have three checks here:

- Length check against zero
- Sortedness check
- Out of bounds check

If a byzantine node sends a block with height 0 and 1,000,000 transactions, this code will attempt to determine if the 1 million transactions are sorted, and then it will check the out of bounds later.

We shouuld always check out of bounds conditions **first** before any expensive operation.